### PR TITLE
Add options to specify number densities as formula units / Ångström^3

### LINKED ISSUE #21468: SetSampleMaterial: no possibility to set number density for formula units
Internally Mantid uses atom number densities. This is also reflected in [`SetSampleMaterial`](http://docs.mantidproject.org/nightly/algorithms/SetSampleMaterial-v1.html) where the input property *SampleNumberDensity* expects the density per atom. In some disciplines, however, the use number densities per formula unit would be more convenient. It should be investigated if an additional input property for the formula unit density could be added to `SetSampleMaterial` and the density converted to the Mantid convention using the *ChemicalFormula* information within the algorithm.

### PR DESCRIPTION
**Pull request by soininen**
Internally we use atoms / Ångström^3 but it is beneficial to permit users to give the number densities as formula units / Ångström^3. This is useful when the chemical formula is complex. Mantid can easily calculate the actual number of atoms and convert to the internal number density representation.

This PR enables the new input unit options in `mantid.kernel.MaterialBuilder`, `SetSampleMaterial` and `LoadSampleEnvironment`.

An additional cleanup considers the return type of some `Material` returning functions which were returning-by-value: these were changed to return-by-reference, as it was clear that these were just missing the ampersand `&`.

**Report to:** Björn Fåk, ILL

**To test:**

The following script should execute without issues:

```
ws = CreateSampleWorkspace()
SetSampleMaterial(
    ws,
    ChemicalFormula='H2 O',
    SampleNumberDensity=0.5,
    NumberDensityUnit='Formula Units')
material = ws.sample().getMaterial()
assert(material.numberDensity, 0.5 * 3)
```

Fixes #21468.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/test/MaterialBuilderTest.h [hunk 1]
```diff
@@ -131,7 +131,7 @@ class MaterialBuilderTest : public CxxTest::TestSuite {
     TS_ASSERT_DELTA(mat.numberDensity(), 0.0410414, 0.001);
   }
 
-  void test_Number_Set_By_AtomicNumber_ZParameter_And_Cell_Volume() {
+  void test_Number_Density_Set_By_AtomicNumber_ZParameter_And_Cell_Volume() {
```
Comment (Pasarus) [inline Framework/Kernel/test/MaterialBuilderTest.h]: Since the rest of the file is in this format for function names, I am not too bothered. But the mantid standards state that:
image
So according to that it should be `test_numberDensitySetByAtomicNumberZParameterAndCellVolume`.

Again the whole file is at fault and that is outside the scope so leave the test names as they are :+1:

File: Framework/PythonInterface/test/python/mantid/kernel/MaterialBuilderTest.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+from __future__ import (absolute_import, division, print_function)
+
+import unittest
+from mantid.kernel import MaterialBuilder, NumberDensityUnit
+
+
+class MaterialBuilderTest(unittest.TestCase):
```
Comment (Pasarus) [inline Framework/PythonInterface/test/python/mantid/kernel/MaterialBuilderTest.py]: :+1:

**Review by Pasarus — CHANGES_REQUESTED** [review]
Besides some minor comments and changes this looks good! I will do a test of the feature now.

**Comment by Pasarus** [comment]
Feature works as expected.

**Comment by Pasarus** [comment]
Also the Windows Failure is a known unreliable test, therefore no need to rebuild and clog the build servers.

File: Framework/DataHandling/inc/MantidDataHandling/ReadMaterial.h [hunk 3]
```diff
@@ -84,8 +87,10 @@ class DLLExport ReadMaterial {
   void setMaterial(const std::string chemicalSymbol, const int atomicNumber,
                    const int massNumber);
 
-  void setNumberDensity(const double rho_m, const double rho,
-                        const double zParameter, const double unitCellVolume);
+  void
+  setNumberDensity(const double rho_m, const double rho,
+                   const Kernel::MaterialBuilder::NumberDensityUnit rhoUnit,
```
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/DataHandling/inc/MantidDataHandling/ReadMaterial.h]: The header files don't have the `using namespace` lines so the `Kernel::` part cannot be avoided.

File: Framework/DataHandling/src/LoadSampleEnvironment.cpp [hunk 4]
```diff
@@ -138,6 +144,9 @@ void LoadSampleEnvironment::init() {
                                             "SetMaterial", IS_NOT_DEFAULT));
   setPropertySettings("SampleMassDensity", make_unique<EnabledWhenProperty>(
                                                "SetMaterial", IS_NOT_DEFAULT));
+  setPropertySettings("NumberDensityUnit",
+                      make_unique<Kernel::EnabledWhenProperty>(
```
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/DataHandling/src/LoadSampleEnvironment.cpp]: I cleaned up `API::` and `Geometry::` namespaces here too.

File: Framework/Kernel/test/MaterialBuilderTest.h (hunk 1 above)
Comment (soininen) [reply Framework/Kernel/test/MaterialBuilderTest.h]: Our documentation for unit test practices has still another style without the underscore:
http://developer.mantidproject.org/UnitTestGoodPractice.html

Traditionally one mixes all possible naming styles within the unit test files.

File: Framework/Geometry/src/Objects/CSGObject.cpp [hunk 5]
```diff
@@ -115,11 +115,11 @@ void CSGObject::setMaterial(const Kernel::Material &material) {
...
  * @return The Material that the object is composed from
  */
-const Kernel::Material CSGObject::material() const {
-  if (m_material)
-    return *m_material;
-  else
-    return Material();
+const Kernel::Material &CSGObject::material() const {
+  if (!m_material) {
+    m_material = Kernel::make_unique<Material>();
```
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/Geometry/src/Objects/CSGObject.cpp]: These old files are actually less scary to modify than new ones since they are covered by lots of unit, doc and system tests at different levels.

The original intention regarding the `material()` function is to clearly return a `const` reference. However, this would return a dangling reference the way `CSGObject::material()` was implemented previously, thus somebody removed the `&` without a second thought.

I believe the reason `CSGObject` uses a `unique_ptr` is to save some memory as there can be a hundred thousand `CSGObject` detectors without material and a single sample shape with one. For some reason, `MeshObject` and `MeshObject2D` went a different route.

Anyway, I think this change gives us the originally intended behavior --- returning the material object as `const` reference.

File: Framework/DataHandling/inc/MantidDataHandling/ReadMaterial.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pasarus) [reply Framework/DataHandling/inc/MantidDataHandling/ReadMaterial.h]: It was my understanding that if you are already in a namespace you do not need to define it. Like how you do not define `Mantid::Kernel` and just `Kernel::`

image

File: Framework/DataHandling/src/LoadSampleEnvironment.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Pasarus) [reply Framework/DataHandling/src/LoadSampleEnvironment.cpp]: Ah good spot :+1:

File: Framework/Geometry/src/Objects/CSGObject.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Pasarus) [reply Framework/Geometry/src/Objects/CSGObject.cpp]: I was speaking with @gemmaguest and since it is a const method, should it really be making a change to the object?

I agree that it could be the expected behavior, however since it is empty by default surely that may have been for a reason, so should it not just be set then? Thus negating the point of this discussion.

File: Framework/Kernel/test/MaterialBuilderTest.h (hunk 1 above)
Comment (Pasarus) [reply Framework/Kernel/test/MaterialBuilderTest.h]: I was following the example given in our standards:
http://developer.mantidproject.org/Standards/UnitTestStandards.html

e.g.:
image

File: Framework/DataHandling/inc/MantidDataHandling/ReadMaterial.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/DataHandling/inc/MantidDataHandling/ReadMaterial.h]: The namespace here is `Mantid::DataHandling` not `Mantid::Kernel`. The `Mantid::Kernel` namespace only extends over the forward declaration of `Material`.

[the remark this answers is not in the document]
Comment (Pasarus) [reply Framework/DataHandling/inc/MantidDataHandling/ReadMaterial.h]: Ok I See now my mistake.

File: Framework/Geometry/src/Objects/CSGObject.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/Geometry/src/Objects/CSGObject.cpp]: From a user's point of view (looking at this class as a black box):
* If no material is set this method returns a `const` reference to an empty material. No modifications to the `CSGObject` object.
* If material is set this method returns a `const` reference to the material. No modifications to the `CSGObject` object.

This is exactly the previous behavior with the exception that now we return a `const` reference to `Material`, not a copy.

[the remark this answers is not in the document]
Comment (Pasarus) [reply Framework/Geometry/src/Objects/CSGObject.cpp]: I see your point!

**Review by Pasarus — APPROVED** [review]

**Comment by NickDraper** [comment]
@soininen Antti, if you can fix the release notes conflict then we can merge this.  Ping me after you've made the change, there is no need to wait for the builds again.

**Comment by soininen** [comment]
@NickDraper The conflicts are gone now. Ping.
